src/to-be-visible.js

```diff
@@ -1,33 +1,63 @@
 import {checkHtmlElement} from './utils'
 
-function isStyleVisible(element) {
+function getElementVisibilityStyle(element) {
+  if (!element) return 'visible'
   const {getComputedStyle} = element.ownerDocument.defaultView
+  const {visibility} = getComputedStyle(element)
+  return visibility || getElementVisibilityStyle(element.parentElement)
+}
 
-  const {display, visibility, opacity} = getComputedStyle(element)
-  return (
-    display !== 'none' &&
-    visibility !== 'hidden' &&
-    visibility !== 'collapse' &&
-    opacity !== '0' &&
-    opacity !== 0
-  )
+function isVisibleSummaryDetails(element, previousElement) {
+  return element.nodeName === 'DETAILS' &&
+    previousElement.nodeName !== 'SUMMARY'
+    ? element.hasAttribute('open')
+    : true
 }
 
-function isAttributeVisible(element, previousElement) {
+function isElementTreeVisible(element, previousElement = undefined) {
+  const {getComputedStyle} = element.ownerDocument.defaultView
+  const {display, opacity} = getComputedStyle(element)
   return (
+    display !== 'none' &&
+    opacity !== '0' &&
     !element.hasAttribute('hidden') &&
-    (element.nodeName === 'DETAILS' && previousElement.nodeName !== 'SUMMARY'
-      ? element.hasAttribute('open')
+    isVisibleSummaryDetails(element, previousElement) &&
+    (element.parentElement
+      ? isElementTreeVisible(element.parentElement, element)
       : true)
   )
 }
 
-function isElementVisible(element, previousElement) {
-  return (
-    isStyleVisible(element) &&
-    isAttributeVisible(element, previousElement) &&
-    (!element.parentElement || isElementVisible(element.parentElement, element))
-  )
+function isElementVisibilityVisible(element) {
+  const visibility = getElementVisibilityStyle(element)
+  return visibility !== 'hidden' && visibility !== 'collapse'
+}
+
+/**
+ * Computes the boolean value that determines if an element is considered visible from the
+ * `toBeVisible` custom matcher point of view.
+ *
+ * Visibility is controlled via two different sets of properties and styles.
+ *
+ * 1. One set of properties allow parent elements to fully controls its sub-tree visibility. This
+ *    means that if higher up in the tree some element is not visible by this criteria, it makes the
+ *    entire sub-tree not visible too, and there's nothing that child elements can do to revert it.
+ *    This includes `display: none`, `opacity: 0`, the presence of the `hidden` attribute`, and the
+ *    open state of a details/summary elements pair.
+ *
+ * 2. The other aspect influencing if an element is visible is the CSS `visibility` style. This one
```

Comment:
nit:

```suggestion
 * 2. The other aspect influencing if an element is visible is the CSS `visibility` property. This one
```
That's how these are named officially i.e. in the spec.